carousel/core/__init__.py

```diff
@@ -243,27 +258,86 @@ class ExampleBase(CommonBase):
+        meta_attrs = get_public_attributes(meta)
+        # check all bases for meta
+        for base in bases:
+            base_meta = getattr(base, mcs._meta_cls, None)
+            # skip if base has no meta
+            if base_meta is None:
+                continue
+            # loop over base meta
+            for a in get_public_attributes(base_meta, as_list=False):
+                # skip if already in meta
+                if a in meta_attrs:
+                    continue
+                # copy meta-option attribute from base
+                setattr(meta, a, getattr(base_meta, a))
+        attr[mcs._meta_attr] = meta  # set _meta combined from bases
+        return attr
 
     @classmethod
     def set_param_file_or_parameters(mcs, attr):
+        """
+        Set parameters from class attributes that are instances of
+        :class:`~carousel.core.Parameter` or from a parameter file.
 
-        meta = attr.pop('Meta', None)
-        if meta is not None:
-            cls_path = getattr(meta, mcs._path_attr, None)
-            cls_file = getattr(meta, mcs._file_attr, None)
-            attr['_meta'] = meta
-        else:
-            cls_path = cls_file = None
+        Any class attributes that are instances of
+        :class:`~carousel.core.Parameter` are popped from the class and added to
+        a the ``parameters`` attribute, which is a dictionary of the parameters.
 
-        # TODO: read parameters from param_file and then also update from attr
+        :param attr: class attributes
+        :return: new list of class attributes with parameters
+        """
+        meta = attr[mcs._meta_attr]  # look for parameter file path in meta
+        cls_path = getattr(meta, mcs._path_attr, None)
+        cls_file = getattr(meta, mcs._file_attr, None)
+        # read parameters
+        attr[mcs._param_attr] = {}
+        attr[mcs._param_file] = None
+        # read parameters from file
         if None not in [cls_path, cls_file]:
-            attr['param_file'] = os.path.join(cls_path, cls_file)
-        else:
-            attr['parameters'] = dict.fromkeys(
-                k for k, v in attr.iteritems() if isinstance(v, Parameter)
-            )
-            for k in attr['parameters']:
-                attr['parameters'][k] = attr.pop(k)
+            param_file = os.path.join(cls_path, cls_file)
+            attr[mcs._param_file] = param_file
+            # read and load JSON parameter map file as "parameters"
+            with open(param_file, 'r') as param_file:
+                file_params = json.load(param_file)
+            # update meta from file
+            for k, v in file_params.pop(mcs._meta_cls, {}).iteritems():
+                setattr(meta, k, v)
+            # dictionary of parameters for reading source file
+            attr[mcs._param_attr] = {
+                k: mcs._param_cls(**v) for k, v in file_params.iteritems()
+            }
+        # get parameters from class
+        parameters = dict.fromkeys(
```

Comment:
not sure if that will work since `attr` is being changed will it is being iterated, maybe it will. Should work if we used [`viewitems()`](https://docs.python.org/2/library/stdtypes.html#dict.viewitems), since that's a view to the items. For [`iteritems()`](https://docs.python.org/2/library/stdtypes.html#dict.iteritems) Python gives this cryptic warning:

>Using iteritems() while adding or deleting entries in the dictionary may raise a RuntimeError or fail to iterate over all entries.

Generally I don't like to change the iterant while it is iterating, but this is probably silly. I think a view should work here. Should I change it?